packages/core/src/odata/v4/filter-function.ts

```diff
@@ -46,15 +73,51 @@ export function filterFunction<EntityT extends Entity>(
  */
 export function filterFunction<EntityT extends Entity>(
   functionName: string,
-  returnType: FilterFunctionReturnType | 'datetimeoffset',
+  returnType: FilterFunctionReturnTypeV4,
   ...parameters: FilterFunctionParameterType<EntityT>[]
 ):
   | BooleanFilterFunction<EntityT>
   | NumberFilterFunction<EntityT>
   | StringFilterFunction<EntityT>
-  | DateFilterFunction<EntityT> {
+  | DateFilterFunction<EntityT>
+  | CollectionFilterFunction<EntityT, any> {
   if (returnType === 'datetimeoffset') {
     return new DateFilterFunction(functionName, parameters);
   }
+  if (isCollectionReturnType(returnType)) {
+    const edmType = returnTypeMapping[returnType.replace('[]', '')];
+    if (edmType) {
+      return new CollectionFilterFunction(functionName, parameters, edmType);
+    }
+    throw new Error(
+      `Cannot create filter function for unknown return type ${returnType}.`
+    );
+  }
   return createFilterFunction(functionName, returnType, ...parameters);
 }
+
+function isCollectionReturnType(
+  returnType: FilterFunctionReturnTypeV4
+): returnType is CollectionReturnType {
+  return returnType.endsWith('[]');
+}
+
+type CollectionReturnType =
+  | 'datetimeoffset[]'
+  | 'boolean[]'
+  | 'int[]'
+  | 'double[]'
+  | 'decimal[]'
+  | 'string[]';
+
+type FilterFunctionReturnTypeV4 =
+  | FilterFunctionReturnType
+  | CollectionReturnType
+  | 'datetimeoffset';
+
+const returnTypeMapping = {
+  datetimeoffset: 'Edm.DateTimeOffset',
+  boolean: 'Edm.Boolean',
+  string: 'Edm.String',
```

Comment:
I was thinking about that initially, but there we only use the numeric type, here we use all for the collection types which are not there in the common filter functions. I can of course move it down, but at the end this will be used only here anyways. Just let me know what you think ;)

Reply:
Ok, because the [boolean](https://github.com/SAP/cloud-sdk/blob/master/packages/core/src/odata/common/filter/filter-function.ts#L61) calls a smarter [constructor](https://github.com/SAP/cloud-sdk/blob/master/packages/core/src/odata/common/filter/boolean-filter-function.ts#L23) so that you don't need the edm type. Therefore, the mapping is useless there.
Feel free to decide then.

Reply:
I'll leave it for now as this way, this is not exported. We might want to consider whether we really need these "smarter" filter function classes, then we will need that in common as well.